coalib/core/Core.py

```diff
@@ -235,11 +237,18 @@ def _schedule_bears(self, bears):
                 logging.debug('Scheduled {!r} (tasks: {})'.format(bear,
                                                                   len(tasks)))
 
+                # Cleanup bears without tasks after all bears had the chance to
+                # schedule their tasks. Not doing so might stop the run too
+                # early, as the cleanup is also responsible for stopping the
+                # event-loop when no more tasks do exist.
                 if not tasks:
-                    # We need to recheck our runtime if something is left to
-                    # process, as when no tasks were offloaded the event-loop
-                    # could hang up otherwise.
-                    self._cleanup_bear(bear)
+                    bears_without_tasks_to_cleanup.append(bear)
+
+        for bear in bears_without_tasks_to_cleanup:
+            # We need to recheck our runtime if something is left to process, as
```

Comment:
Do we still need this comment ?

Reply:
hm yeah... I think this is more or less self explaining :+1: